crates/accelerate/src/circuit_library/utils.rs

```diff
@@ -0,0 +1,58 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use qiskit_circuit::{
+    imports,
+    operations::{Param, PyInstruction},
+    packed_instruction::PackedOperation,
+    Clbit, Qubit,
+};
+use smallvec::{smallvec, SmallVec};
+
+type Instruction = (
+    PackedOperation,
+    SmallVec<[Param; 3]>,
+    Vec<Qubit>,
+    Vec<Clbit>,
+);
+
+/// Get an iterator that returns a barrier or an empty element.
+pub fn maybe_barrier(
+    py: Python,
+    num_qubits: u32,
+    insert_barriers: bool,
+) -> Box<dyn Iterator<Item = PyResult<Instruction>>> {
+    // TODO could speed this up by only defining the barrier class once
+    if !insert_barriers {
+        Box::new(std::iter::empty())
+    } else {
+        let barrier_cls = imports::BARRIER.get_bound(py);
+        let barrier = barrier_cls
+            .call1((num_qubits,))
+            .expect("Could not create Barrier Python-side");
+        let barrier_inst = PyInstruction {
+            qubits: num_qubits,
+            clbits: 0,
+            params: 0,
+            op_name: "barrier".to_string(),
+            control_flow: false,
+            instruction: barrier.into(),
+        };
+        Box::new(std::iter::once(Ok((
+            barrier_inst.into(),
+            smallvec![],
+            (0..num_qubits).map(Qubit).collect(),
+            vec![] as Vec<Clbit>,
+        ))))
+    }
+}
```

Comment:
I can put `then_some` but I think this has one disadvantage: when we write something like
```rust
instructions.chain(
   insert_barrier.then_some(Ok(barrier.clone())
)
```
Then even if `insert_barrier` is `false`, I need to make the Python call to construct the barrier object. To avoid that, what I would like to do is 
```rust
let barrier = match insert_barrier {
    true => Some(.... get barrier instruction...),
    false => None,
}

// in my iteration loop ...
    instructions.chain(
        insert_barrier.then_some(Ok(barrier.clone().unwrap()))
    )
```
but rust won't allow this since `barrier` can be `None` and I'm not allowed to unwrap 🤔